experimental/lib/Analysis/GsaAnalysis.cpp

```diff
@@ -0,0 +1,485 @@
+  struct MissingPhi {
+
+    // Which input is missing
+    GateInput *pi;
+
+    // Related block argument
+    BlockArgument ba;
+
+    MissingPhi(GateInput *pi, BlockArgument ba) : pi(pi), ba(ba) {}
+  };
+
+  // Initialize the index of the class
+  uniqueGateIndex = 0;
+
+  // Vector to store all the missing phi
+  SmallVector<MissingPhi> phisToConnect;
+
+  // For each block in the function
+  for (Block &block : funcOp.getBlocks()) {
+
+    // Create an empty list for the phi functions corresponding to the block
+    gateList.insert({&block, llvm::SmallVector<Gate *>()});
+
+    // For each block argument
+    for (BlockArgument &arg : block.getArguments()) {
+      unsigned argNumber = arg.getArgNumber();
+      // Create a set for the operands of the corresponding phi function
+      SmallVector<GateInput *> operands;
+      DenseSet<Block *> coveredPredecessors;
+      // For each predecessor of the block, which is in charge of
+      // providing the inputs of the phi functions
+      for (Block *pred : block.getPredecessors()) {
+
+        // Make sure that a predecessor is covered only once
+        if (coveredPredecessors.contains(pred))
+          continue;
+        coveredPredecessors.insert(pred);
+
+        // Get the branch terminator
+        auto branchOp = dyn_cast<BranchOpInterface>(pred->getTerminator());
+        assert(branchOp && "Expected terminator operation in a predecessor "
+                           "block feeding a block argument!");
+
+        // Check if the input value `c` of type Value is already present among
+        // the operands of the phi function
+        auto isAlreadyPresent = [&](Value c) -> bool {
+          return std::any_of(operands.begin(), operands.end(),
+                             [c](GateInput *in) {
+                               return in->isTypeValue() && in->getValue() == c;
+                             });
+        };
+
+        // For each alternative in the branch terminator
+        for (auto [successorId, successorBlock] :
+             llvm::enumerate(branchOp->getSuccessors())) {
+
+          // Skip the successor if it is not the block under analysis
+          if (successorBlock != &block)
+            continue;
+
+          // Get the values used for that branch
+          auto successorOperands = branchOp.getSuccessorOperands(successorId);
+          // Get the value used as input of the gate
+          auto producer = successorOperands[argNumber];
+          GateInput *gateInput = nullptr;
+
+          /// If it is a block argument whose parent block has some predecessor,
+          /// then the value is the output of a phi, and we add it to the list
+          /// of missing phis. Otherwise, the input is a value, and it can be
+          /// safely added directly.
+          if (auto ba = dyn_cast<BlockArgument>(producer);
+              ba && !producer.getParentBlock()->hasNoPredecessors()) {
+            gateInput = new GateInput((Gate *)nullptr);
```

Comment:
Where is this memory deallocated? The analysis should take care of that in its destructor most likely.